alf/algorithms/sac_algorithm.py

```diff
@@ -290,32 +287,71 @@ def actor_loss_fn(dqda, action):
             log_action_probs = base_action_dist.logits.squeeze(1)
 
             target_q_value = critics.min(dim=1)[0].detach()
-            alpha = torch.exp(self._log_alpha)
             actor_loss = torch.exp(log_action_probs) * (
-                alpha.detach() * log_action_probs - target_q_value)
+                alpha * log_action_probs - target_q_value)
             actor_loss = actor_loss.mean(list(range(1, actor_loss.ndim)))
+        else:
+            # With CriticNetwork, we can only make one action's probability
+            # proportional to the corresponding Q value at a time.
+            # If continuous, use the original SAC's formulation; otherwise only
+            # train one entry of a categorical distribution.
+            # In either case, the target Q value to be fit to is the same.
+            target_q_value = critics.min(dim=1)[0]
+            cont_act_logpi = nest.flatten_up_to(
+                self._is_continuous,
+                nest.map_structure(lambda *args: tuple(args),
+                                   self._is_continuous, action, log_pi))
+
+            def _train_continuous_actor(action, log_pi):
+                dqda = list(torch.autograd.grad(target_q_value.sum(), action))
+
+                def actor_loss_fn(dqda, action, log_pi):
+                    if self._dqda_clipping:
+                        dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                           self._dqda_clipping)
+                    loss = 0.5 * losses.element_wise_squared_loss(
+                        (dqda + action).detach(), action)
+                    loss = loss.sum(list(range(1, loss.ndim)))
+                    loss += alpha * log_pi
+                    return loss
+
+                return nest.map_structure(actor_loss_fn, dqda, action, log_pi)
+
+            def _train_discrete_actor(log_pi):
+                def actor_loss_fn(log_pi):
+                    # See ``notes/estimating_derivative_of_expectation.rst``
+                    # for how this loss formula is derived.
+                    loss = log_pi * (alpha * log_pi - target_q_value).detach()
```

Comment:
On carpole, this version will converge slower than using a QNetwork, but eventually got 200. 

This version:
![Screen Shot 2020-04-28 at 11 35 31 AM](https://user-images.githubusercontent.com/51248379/80524534-a0655b80-8944-11ea-9759-943a0866ea0d.png)

Previous:
![Screen Shot 2020-04-28 at 11 21 16 AM](https://user-images.githubusercontent.com/51248379/80524554-a8250000-8944-11ea-9a38-7f2fedf5d87a.png)

Reply:
To handle a mixed action case, it seems possible to use the original critic network for continuous part of the actions, and the q-network for discrete part of the actions,  if that is beneficial for the final performance?

Reply:
If the critic network and q-network are separate, then learning either of them will be non-stationary because the reward is decided by both actions. It might work if these two networks share some early layers, but it seems to me that the sharing is unintuitive.